// three/src/materials/LineBasicMaterial.js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { Material } from './Material.js';
import { Color } from '../math/Color.js';
/**
 * parameters = {
 *  color: <hex>,
 *  opacity: <float>,
 *
 *  linewidth: <float>,
 *  linecap: "round",
 *  linejoin: "round"
 * }
 */

var LineBasicMaterial = /*#__PURE__*/function (_Material) {
  _inheritsLoose(LineBasicMaterial, _Material);

  function LineBasicMaterial(parameters) {
    var _this;

    _this = _Material.call(this) || this;
    _this.type = 'LineBasicMaterial';
    _this.color = new Color(0xffffff);
    _this.linewidth = 1;
    _this.linecap = 'round';
    _this.linejoin = 'round';

    _this.setValues(parameters);

    return _this;
  }

  var _proto = LineBasicMaterial.prototype;

  _proto.copy = function copy(source) {
    _Material.prototype.copy.call(this, source);

    this.color.copy(source.color);
    this.linewidth = source.linewidth;
    this.linecap = source.linecap;
    this.linejoin = source.linejoin;
    return this;
  };

  return LineBasicMaterial;
}(Material);

LineBasicMaterial.prototype.isLineBasicMaterial = true;
export { LineBasicMaterial };
