// three/src/materials/MeshMatcapMaterial.js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { TangentSpaceNormalMap } from '../constants.js';
import { Material } from './Material.js';
import { Vector2 } from '../math/Vector2.js';
import { Color } from '../math/Color.js';
/**
 * parameters = {
 *  color: <hex>,
 *  opacity: <float>,
 *
 *  matcap: new THREE.Texture( <Image> ),
 *
 *  map: new THREE.Texture( <Image> ),
 *
 *  bumpMap: new THREE.Texture( <Image> ),
 *  bumpScale: <float>,
 *
 *  normalMap: new THREE.Texture( <Image> ),
 *  normalMapType: THREE.TangentSpaceNormalMap,
 *  normalScale: <Vector2>,
 *
 *  displacementMap: new THREE.Texture( <Image> ),
 *  displacementScale: <float>,
 *  displacementBias: <float>,
 *
 *  alphaMap: new THREE.Texture( <Image> ),
 *
 *  flatShading: <bool>
 * }
 */

var MeshMatcapMaterial = /*#__PURE__*/function (_Material) {
  _inheritsLoose(MeshMatcapMaterial, _Material);

  function MeshMatcapMaterial(parameters) {
    var _this;

    _this = _Material.call(this) || this;
    _this.defines = {
      'MATCAP': ''
    };
    _this.type = 'MeshMatcapMaterial';
    _this.color = new Color(0xffffff); // diffuse

    _this.matcap = null;
    _this.map = null;
    _this.bumpMap = null;
    _this.bumpScale = 1;
    _this.normalMap = null;
    _this.normalMapType = TangentSpaceNormalMap;
    _this.normalScale = new Vector2(1, 1);
    _this.displacementMap = null;
    _this.displacementScale = 1;
    _this.displacementBias = 0;
    _this.alphaMap = null;
    _this.flatShading = false;

    _this.setValues(parameters);

    return _this;
  }

  var _proto = MeshMatcapMaterial.prototype;

  _proto.copy = function copy(source) {
    _Material.prototype.copy.call(this, source);

    this.defines = {
      'MATCAP': ''
    };
    this.color.copy(source.color);
    this.matcap = source.matcap;
    this.map = source.map;
    this.bumpMap = source.bumpMap;
    this.bumpScale = source.bumpScale;
    this.normalMap = source.normalMap;
    this.normalMapType = source.normalMapType;
    this.normalScale.copy(source.normalScale);
    this.displacementMap = source.displacementMap;
    this.displacementScale = source.displacementScale;
    this.displacementBias = source.displacementBias;
    this.alphaMap = source.alphaMap;
    this.flatShading = source.flatShading;
    return this;
  };

  return MeshMatcapMaterial;
}(Material);

MeshMatcapMaterial.prototype.isMeshMatcapMaterial = true;
export { MeshMatcapMaterial };
