// three/src/materials/MeshNormalMaterial.js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { TangentSpaceNormalMap } from '../constants.js';
import { Material } from './Material.js';
import { Vector2 } from '../math/Vector2.js';
/**
 * parameters = {
 *  opacity: <float>,
 *
 *  bumpMap: new THREE.Texture( <Image> ),
 *  bumpScale: <float>,
 *
 *  normalMap: new THREE.Texture( <Image> ),
 *  normalMapType: THREE.TangentSpaceNormalMap,
 *  normalScale: <Vector2>,
 *
 *  displacementMap: new THREE.Texture( <Image> ),
 *  displacementScale: <float>,
 *  displacementBias: <float>,
 *
 *  wireframe: <boolean>,
 *  wireframeLinewidth: <float>
 *
 *  flatShading: <bool>
 * }
 */

var MeshNormalMaterial = /*#__PURE__*/function (_Material) {
  _inheritsLoose(MeshNormalMaterial, _Material);

  function MeshNormalMaterial(parameters) {
    var _this;

    _this = _Material.call(this) || this;
    _this.type = 'MeshNormalMaterial';
    _this.bumpMap = null;
    _this.bumpScale = 1;
    _this.normalMap = null;
    _this.normalMapType = TangentSpaceNormalMap;
    _this.normalScale = new Vector2(1, 1);
    _this.displacementMap = null;
    _this.displacementScale = 1;
    _this.displacementBias = 0;
    _this.wireframe = false;
    _this.wireframeLinewidth = 1;
    _this.fog = false;
    _this.flatShading = false;

    _this.setValues(parameters);

    return _this;
  }

  var _proto = MeshNormalMaterial.prototype;

  _proto.copy = function copy(source) {
    _Material.prototype.copy.call(this, source);

    this.bumpMap = source.bumpMap;
    this.bumpScale = source.bumpScale;
    this.normalMap = source.normalMap;
    this.normalMapType = source.normalMapType;
    this.normalScale.copy(source.normalScale);
    this.displacementMap = source.displacementMap;
    this.displacementScale = source.displacementScale;
    this.displacementBias = source.displacementBias;
    this.wireframe = source.wireframe;
    this.wireframeLinewidth = source.wireframeLinewidth;
    this.flatShading = source.flatShading;
    return this;
  };

  return MeshNormalMaterial;
}(Material);

MeshNormalMaterial.prototype.isMeshNormalMaterial = true;
export { MeshNormalMaterial };
