// three/src/materials/MeshDepthMaterial.js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { Material } from './Material.js';
import { BasicDepthPacking } from '../constants.js';
/**
 * parameters = {
 *
 *  opacity: <float>,
 *
 *  map: new THREE.Texture( <Image> ),
 *
 *  alphaMap: new THREE.Texture( <Image> ),
 *
 *  displacementMap: new THREE.Texture( <Image> ),
 *  displacementScale: <float>,
 *  displacementBias: <float>,
 *
 *  wireframe: <boolean>,
 *  wireframeLinewidth: <float>
 * }
 */

var MeshDepthMaterial = /*#__PURE__*/function (_Material) {
  _inheritsLoose(MeshDepthMaterial, _Material);

  function MeshDepthMaterial(parameters) {
    var _this;

    _this = _Material.call(this) || this;
    _this.type = 'MeshDepthMaterial';
    _this.depthPacking = BasicDepthPacking;
    _this.map = null;
    _this.alphaMap = null;
    _this.displacementMap = null;
    _this.displacementScale = 1;
    _this.displacementBias = 0;
    _this.wireframe = false;
    _this.wireframeLinewidth = 1;
    _this.fog = false;

    _this.setValues(parameters);

    return _this;
  }

  var _proto = MeshDepthMaterial.prototype;

  _proto.copy = function copy(source) {
    _Material.prototype.copy.call(this, source);

    this.depthPacking = source.depthPacking;
    this.map = source.map;
    this.alphaMap = source.alphaMap;
    this.displacementMap = source.displacementMap;
    this.displacementScale = source.displacementScale;
    this.displacementBias = source.displacementBias;
    this.wireframe = source.wireframe;
    this.wireframeLinewidth = source.wireframeLinewidth;
    return this;
  };

  return MeshDepthMaterial;
}(Material);

MeshDepthMaterial.prototype.isMeshDepthMaterial = true;
export { MeshDepthMaterial };
