// three/src/materials/PointsMaterial.js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { Material } from './Material.js';
import { Color } from '../math/Color.js';
/**
 * parameters = {
 *  color: <hex>,
 *  opacity: <float>,
 *  map: new THREE.Texture( <Image> ),
 *  alphaMap: new THREE.Texture( <Image> ),
 *
 *  size: <float>,
 *  sizeAttenuation: <bool>
 *
 * }
 */

var PointsMaterial = /*#__PURE__*/function (_Material) {
  _inheritsLoose(PointsMaterial, _Material);

  function PointsMaterial(parameters) {
    var _this;

    _this = _Material.call(this) || this;
    _this.type = 'PointsMaterial';
    _this.color = new Color(0xffffff);
    _this.map = null;
    _this.alphaMap = null;
    _this.size = 1;
    _this.sizeAttenuation = true;

    _this.setValues(parameters);

    return _this;
  }

  var _proto = PointsMaterial.prototype;

  _proto.copy = function copy(source) {
    _Material.prototype.copy.call(this, source);

    this.color.copy(source.color);
    this.map = source.map;
    this.alphaMap = source.alphaMap;
    this.size = source.size;
    this.sizeAttenuation = source.sizeAttenuation;
    return this;
  };

  return PointsMaterial;
}(Material);

PointsMaterial.prototype.isPointsMaterial = true;
export { PointsMaterial };
