// three/src/loaders/Loader.js
import { DefaultLoadingManager } from './LoadingManager.js';

var Loader = /*#__PURE__*/function () {
  function Loader(manager) {
    this.manager = manager !== undefined ? manager : DefaultLoadingManager;
    this.crossOrigin = 'anonymous';
    this.withCredentials = false;
    this.path = '';
    this.resourcePath = '';
    this.requestHeader = {};
  }

  var _proto = Loader.prototype;

  _proto.load = function
  /* url, onLoad, onProgress, onError */
  load() {};

  _proto.loadAsync = function loadAsync(url, onProgress) {
    var scope = this;
    return new Promise(function (resolve, reject) {
      scope.load(url, resolve, onProgress, reject);
    });
  };

  _proto.parse = function
  /* data */
  parse() {};

  _proto.setCrossOrigin = function setCrossOrigin(crossOrigin) {
    this.crossOrigin = crossOrigin;
    return this;
  };

  _proto.setWithCredentials = function setWithCredentials(value) {
    this.withCredentials = value;
    return this;
  };

  _proto.setPath = function setPath(path) {
    this.path = path;
    return this;
  };

  _proto.setResourcePath = function setResourcePath(resourcePath) {
    this.resourcePath = resourcePath;
    return this;
  };

  _proto.setRequestHeader = function setRequestHeader(requestHeader) {
    this.requestHeader = requestHeader;
    return this;
  };

  return Loader;
}();

export { Loader };
