// three/src/core/Clock.js
var Clock = /*#__PURE__*/function () {
  function Clock(autoStart) {
    if (autoStart === void 0) {
      autoStart = true;
    }

    this.autoStart = autoStart;
    this.startTime = 0;
    this.oldTime = 0;
    this.elapsedTime = 0;
    this.running = false;
  }

  var _proto = Clock.prototype;

  _proto.start = function start() {
    this.startTime = now();
    this.oldTime = this.startTime;
    this.elapsedTime = 0;
    this.running = true;
  };

  _proto.stop = function stop() {
    this.getElapsedTime();
    this.running = false;
    this.autoStart = false;
  };

  _proto.getElapsedTime = function getElapsedTime() {
    this.getDelta();
    return this.elapsedTime;
  };

  _proto.getDelta = function getDelta() {
    var diff = 0;

    if (this.autoStart && !this.running) {
      this.start();
      return 0;
    }

    if (this.running) {
      var newTime = now();
      diff = (newTime - this.oldTime) / 1000;
      this.oldTime = newTime;
      this.elapsedTime += diff;
    }

    return diff;
  };

  return Clock;
}();

function now() {
  return (typeof performance === 'undefined' ? Date : performance).now(); // see #10732
}

export { Clock };
